trin-core/src/portalnet/utp.rs

```diff
@@ -0,0 +1,1022 @@
+        assert_eq!(packet.version(), VERSION);
+        assert_eq!(packet.extension(), 1);
+        assert_eq!(packet.connection_id(), 42054);
+        assert_eq!(packet.timestamp(), 2805920832);
+        assert_eq!(packet.timestamp_difference(), 10000);
+        assert_eq!(packet.wnd_size(), 61440);
+        assert_eq!(packet.seq_nr(), 15404);
+        assert_eq!(packet.ack_nr(), 32437);
+        assert!(packet.get_payload().is_empty());
+    }
+
+    #[test]
+    fn test_decode_packet_with_invalid_extension() {
+        let buf = [
+            0x21, 0x1, 0xA4, 0x46, 0xA7, 0x3E, 0xF4, 0x40, 0x0, 0x0, 0x27, 0x10, 0x0, 0x0, 0xF0,
+            0x0, 0x3C, 0x2C, 0x7E, 0xB5, 0x0, 0x4, 0xAA, 0x3C, 0x0,
+        ];
+
+        let packet = Packet::try_from(&buf[..]);
+        assert!(packet.is_err());
+    }
+
+    #[test]
+    fn test_encode_packet() {
+        let mut response = PacketHeader::new();
+        response.set_type(Type::StData);
+        response.set_connection_id(49300);
+        response.set_timestamp(2805920832);
+        response.set_timestamp_difference(1805367832);
+        response.set_wnd_size(61440);
+        response.set_seq_nr(12044);
+        response.set_ack_nr(12024);
+
+        let packet = Packet::new(response);
+        assert_eq!(packet.type_(), Type::StData);
+        assert_eq!(packet.version(), VERSION);
+        assert_eq!(packet.extension(), 0);
+        assert_eq!(packet.connection_id(), 49300);
+        assert_eq!(packet.timestamp(), 2805920832);
+        assert_eq!(packet.timestamp_difference(), 1805367832);
+        assert_eq!(packet.wnd_size(), 61440);
+        assert_eq!(packet.seq_nr(), 12044);
+        assert_eq!(packet.ack_nr(), 12024);
+        assert!(packet.get_payload().is_empty());
+    }
+
+    #[test]
+    fn test_encode_packet_with_payload() {
+        let mut response = PacketHeader::new();
+        response.set_type(Type::StData);
+        response.set_connection_id(49300);
+        response.set_timestamp(2805920832);
+        response.set_timestamp_difference(1805367832);
+        response.set_wnd_size(61440);
+        response.set_seq_nr(12044);
+        response.set_ack_nr(12024);
+        let payload = b"Hello world".to_vec();
+
+        let packet = Packet::with_payload(response, &payload[..]);
+        assert_eq!(packet.type_(), Type::StData);
+        assert_eq!(packet.version(), VERSION);
+        assert_eq!(packet.extension(), 0);
+        assert_eq!(packet.connection_id(), 49300);
+        assert_eq!(packet.timestamp(), 2805920832);
+        assert_eq!(packet.timestamp_difference(), 1805367832);
+        assert_eq!(packet.wnd_size(), 61440);
+        assert_eq!(packet.seq_nr(), 12044);
+        assert_eq!(packet.ack_nr(), 12024);
+        assert_eq!(packet.get_payload(), &payload[..]);
+    }
+
+    #[test]
+    fn test_empty_packet() {
+        let packet = Packet::try_from(&[][..]);
+        assert!(packet.is_err());
+    }
+
+    #[test]
+    fn test_selective_ack() {
```

Comment:
A little information on where this test vector came from would be handy, when someone is invariably debugging something they broke later, and want to understand the test. Either a link to the test if it's explained or inspired by elsewhere, or some commentary about what's happening along the way: what should be happening, and the kinds of things that might go wrong.

Reply:
Just testing to see if they initialize properly and encode/decode properly (I actually fixed a few bugs writing these). I also wrote the selective ack test to just make sure it properly writes it and test things I thought might have thrown an error, luckily it was good and I cleaned up the code a bit from it.

So in essence I just wanted to find bugs xD other then my integration test I didn't add to the PR these were where I thought possible issues could be